tests/unit/browser/webkit/network/test_pac.py

```diff
@@ -0,0 +1,173 @@
+
+from PyQt5.QtCore import QUrl, QT_VERSION_STR
+from PyQt5.QtNetwork import (QNetworkProxy, QNetworkProxyQuery, QHostInfo,
+                             QHostAddress)
+
+from qutebrowser.browser.webkit.network import pac
+
+
+pytestmark = pytest.mark.usefixtures('qapp')
+
+
+def _pac_common_test(test_str):
+    fun_str_f = """
+        function FindProxyForURL(domain, host) {{
+            {}
+            return "DIRECT; PROXY 127.0.0.1:8080; SOCKS 192.168.1.1:4444";
+        }}
+    """
+
+    fun_str = fun_str_f.format(test_str)
+    res = pac.PACResolver(fun_str)
+    proxies = res.resolve(QNetworkProxyQuery(QUrl("https://example.com/test")))
+    assert len(proxies) == 3
+    assert proxies[0].type() == QNetworkProxy.NoProxy
+    assert proxies[1].type() == QNetworkProxy.HttpProxy
+    assert proxies[1].hostName() == "127.0.0.1"
+    assert proxies[1].port() == 8080
+    assert proxies[2].type() == QNetworkProxy.Socks5Proxy
+    assert proxies[2].hostName() == "192.168.1.1"
+    assert proxies[2].port() == 4444
+
+
+def _pac_equality_test(call, expected):
+    test_str_f = """
+        var res = ({0});
+        var expected = ({1});
+        if(res !== expected) {{
+            throw new Error("failed test {0}: got '" + res + "', expected '" + expected + "'");
+        }}
+    """
+    _pac_common_test(test_str_f.format(call, expected))
+
+
+def _pac_except_test(caplog, call):
+    test_str_f = """
+        var thrown = false;
+        try {{
+            var res = ({0});
+        }} catch(e) {{
+            thrown = true;
+        }}
+        if(!thrown) {{
+            throw new Error("failed test {0}: got '" + res + "', expected exception");
+        }}
+    """
+    with caplog.at_level(logging.ERROR):
+        _pac_common_test(test_str_f.format(call))
+
+
+def _pac_noexcept_test(call):
+    test_str_f = """
+        var res = ({0});
+    """
+    _pac_common_test(test_str_f.format(call))
+
+
+# pylint: disable=line-too-long, invalid-name
+
+
+@pytest.mark.parametrize("domain, expected", [
+    ("known.domain", "'1.2.3.4'"),
+    ("bogus.domain.foobar", "null")
+])
+def test_dnsResolve(monkeypatch, domain, expected):
+    def mock_fromName(host):
+        info = QHostInfo()
+        if host == "known.domain":
+            info.setAddresses([QHostAddress("1.2.3.4")])
+        return info
+    monkeypatch.setattr(QHostInfo, 'fromName', mock_fromName)
+    _pac_equality_test("dnsResolve('{}')".format(domain), expected)
+
+
+def test_myIpAddress():
+    _pac_equality_test("isResolvable(myIpAddress())", "true")
+
+
+def test_proxyBindings():
+    _pac_equality_test("JSON.stringify(ProxyConfig.bindings)", "'{}'")
+
+
+def test_invalid_port():
+    test_str = """
+        function FindProxyForURL(domain, host) {
+            return "PROXY 127.0.0.1:FOO";
+        }
+    """
+
+    res = pac.PACResolver(test_str)
+    with pytest.raises(pac.ParseProxyError):
+        res.resolve(QNetworkProxyQuery(QUrl("https://example.com/test")))
+
+
```

Comment:
Ah, right - I don't know really. I first thought it wouldn't be required, but seeing that there are so many versions of this file around, maybe some tests (especially for the corner cases MDN shows) would make sense. I don't really mind much either way though, I'd rather see more Python code covered by tests as a priority.